commands/pull.js

```diff
@@ -0,0 +1,145 @@
+'use strict'
+
+const cli = require('heroku-cli-util')
+const co = require('co')
+const debug = require('debug')('push')
+const psql = require('../lib/psql')
+
+function parseURL (db) {
+  const url = require('url')
+  db = url.parse(db.match(/:\/\//) ? db : `postgres:///${db}`)
+  let [user, password] = db.auth ? db.auth.split(':') : []
+  db.user = user
+  db.password = password
+  db.database = db.path ? db.path.split('/', 2)[1] : null
+  db.host = db.hostname || 'localhost'
+  return db
+}
+
+function exec (cmd, opts = {}) {
+  const {execSync} = require('child_process')
+  debug(cmd)
+  opts = Object.assign({}, opts, {stdio: 'inherit'})
+  try {
+    return execSync(cmd, opts)
+  } catch (err) {
+    if (err.status) process.exit(err.status)
+    throw err
+  }
+}
+
+const prepare = co.wrap(function * (target) {
+  if (target.host === 'localhost') {
+    exec(`createdb ${connstring(target, true)}`)
+  } else {
+    let result = yield psql.exec(target, 'SELECT COUNT(*) = 0 AS empty FROM pg_stat_user_tables')
+    if (!result.startsWith(' empty \n-------\n t')) throw new Error(`Remote database is not empty. Please create a new database or use ${cli.color.cmd('heroku pg:reset')}`)
+  }
+})
+
+function connstring (uri, skipDFlag) {
+  let user = uri.user ? `-U ${uri.user}` : ''
+  return `${user} -h ${uri.host} -p ${uri.port || 5432} ${skipDFlag ? '' : '-d'} ${uri.database}`
+}
+
+const verifyExtensionsMatch = co.wrap(function * (source, target) {
+  // It's pretty common for local DBs to not have extensions available that
+  // are used by the remote app, so take the final precaution of warning if
+  // the extensions available in the local database don't match. We don't
+  // report it if the difference is solely in the version of an extension
+  // used, though.
+  let sql = 'SELECT extname FROM pg_extension ORDER BY extname;'
+  let extensions = yield {
+    target: psql.exec(target, sql),
+    source: psql.exec(source, sql)
+  }
+  if (extensions.target !== extensions.source) {
+    cli.warn(`WARNING: Extensions in newly created target database differ from existing source database.
+Target extensions:
+${extensions.target}
+Source extensions:
+${extensions.source}
+HINT: You should review output to ensure that any errors
+ignored are acceptable - entire tables may have been missed, where a dependency
+could not be resolved. You may need to to install a postgresql-contrib package
+and retry.`)
+  }
+})
+
+const run = co.wrap(function * (source, target) {
+  yield prepare(target)
+  let password = p => p ? ` PGPASSWORD="${p}"` : ''
```

Comment:
This doesn't handle passwords with double quote characters. We don't generate these, but users may be using that locally. I'm guessing this is another legacy thing copied though, we probably don't need to worry about it now.